kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,118 @@ struct LogReplayScanner {
+/// replay visits actions newest-first, so once we've seen a file action for a given (path, dvId)
+/// pair, we should ignore all subsequent (older) actions for that same (path, dvId) pair. If the
+/// first action for a given file is a remove, then that file does not show up in the result at all.
+struct AddRemoveDedupVisitor<'seen> {
+    seen: &'seen mut HashSet<FileActionKey>,
+    selection_vector: Vec<bool>,
     is_log_batch: bool,
 }
 
-const ADD_FIELD_COUNT: usize = 15;
+impl<'seen> AddRemoveDedupVisitor<'seen> {
+    /// Checks if log replay already processed this logical file (in which case the current action
+    /// should be ignored). If not already seen, register it so we can recognize future duplicates.
+    fn already_seen(&mut self, key: FileActionKey) -> bool {
+        // Note: each (add.path + add.dv_unique_id()) pair has a
+        // unique Add + Remove pair in the log. For example:
+        // https://github.com/delta-io/delta/blob/master/spark/src/test/resources/delta/table-with-dv-large/_delta_log/00000000000000000001.json
 
-impl AddRemoveVisitor {
-    fn new(selection_vector: Option<Vec<bool>>, is_log_batch: bool) -> Self {
-        AddRemoveVisitor {
-            selection_vector,
-            is_log_batch,
-            ..Default::default()
+        if self.seen.contains(&key) {
+            debug!(
+                "Ignoring duplicate ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            true
+        } else {
+            debug!(
+                "Including ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            if self.is_log_batch {
+                // Remember file actions from this batch so we can ignore duplicates as we process
+                // batches from older commit and/or checkpoint files. We don't track checkpoint
+                // batches because they are already the oldest actions and never replace anything.
+                self.seen.insert(key);
+            }
+            false
         }
     }
+
+    /// True if this row contains an Add action that should survive log replay. Skip it if the row
+    /// is not an Add action, or the file has already been seen previously.
+    fn is_valid_add<'a>(&mut self, i: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<bool> {
+        // Add will have a path at index 0 if it is valid; otherwise, if it is a log batch, we
+        // may have a remove with a path at index 4.
+        let (path, getters, is_add) = if let Some(path) = getters[0].get_str(i, "add.path")? {
+            (path, &getters[1..4], true)
+        } else if !self.is_log_batch {
+            return Ok(false);
+        } else if let Some(path) = getters[4].get_opt(i, "remove.path")? {
+            (path, &getters[5..8], false)
+        } else {
+            return Ok(false);
+        };
+
```

Comment:
nit: could you add a comment here that now `getters` are the 3 dv fields. Since it has the same name and is being reassigned in a tuple assignment it's easy to miss that it's not the `getters` that was passed to the function. could also just give it a different name like `dv_getters` or something.

Reply:
I like `dv_getters` a lot. Also added the comment.